Interference between CI runs

Consider the following run: https://github.com/runtimeverification/evm-semantics/actions/runs/3468304857/jobs/5794152325

The job fails because `make kevm-pyk` fails. The log indicates that during `poetry install`, several packages are already installed with another version, in particular:

```
  • Updating pyk (0.1.52 08ab1a0 -> 0.1.34 6d12754)
```

This downgrade probably leaves some garbage from `0.1.52`, thus `check-mypy` fails:

```
src/kevm_pyk/utils.py:5: error: Module "pyk.kast" has no attribute "KApply"
```


The root of the issue is probably in `./.github/actions/with-docker`: the container is set up by mounting both `/opt/workspace` and `${HOME}` as volumes. A possible solution is to use `docker cp` instead (an [example](https://github.com/runtimeverification/pyk/blob/2c0ccd010ad6489662794aed8917538b16863464/.github/workflows/test-pr.yml#L74) of how containers are set up in `pyk`).

--- a/.github/actions/with-docker/action.yml
+++ b/.github/actions/with-docker/action.yml
@@ -24,16 +24,16 @@ runs:
         --tag runtimeverification/${TAG_NAME}                      \
         --build-arg K_COMMIT=${K_COMMIT}                           \
         --build-arg Z3_VERSION=${Z3_VERSION}                       \
-        --build-arg USER_ID=$(id -u) --build-arg GROUP_ID=$(id -g)
 
       docker run                        \
         --name ${TAG_NAME}              \
-        --rm -it                        \
+        --rm                            \
+        --interactive                   \
+        --tty                           \
         --detach                        \
-        --workdir /opt/workspace        \
-        --user $(id -u):$(id -g)        \
-        -v ${HOME}:${HOME}              \
-        -v "$(pwd):/opt/workspace"      \
-        -v "/etc/passwd:/etc/passwd:ro" \
-        -v "/etc/group:/etc/group:ro"   \
+        --user root                     \
+        --workdir /home/user            \
         runtimeverification/${TAG_NAME}
+
+      docker cp . ${TAG_NAME}:/home/user
+      docker exec ${TAG_NAME} chown -R user:user /home/user

--- a/.github/workflows/test-pr.yml
+++ b/.github/workflows/test-pr.yml
@@ -23,21 +23,21 @@ jobs:
         with:
           tag: kevm-ci-concrete-${{ github.sha }}
       - name: 'Build kevm-pyk'
-        run: docker exec -t -u user kevm-ci-concrete-${GITHUB_SHA} /bin/bash -c 'make kevm-pyk venv'
+        run: docker exec -u user kevm-ci-concrete-${GITHUB_SHA} /bin/bash -c 'make kevm-pyk venv'
       - name: 'Build kevm'
-        run: docker exec -t -u user kevm-ci-concrete-${GITHUB_SHA} /bin/bash -c 'make build-llvm build-node -j`nproc` RELEASE=true'
+        run: docker exec -u user kevm-ci-concrete-${GITHUB_SHA} /bin/bash -c 'make build-llvm build-node -j`nproc` RELEASE=true'
       - name: 'Test kevm-pyk'
-        run: docker exec -t -u user kevm-ci-concrete-${GITHUB_SHA} /bin/bash -c 'make test-kevm-pyk -j`nproc`'
+        run: docker exec -u user kevm-ci-concrete-${GITHUB_SHA} /bin/bash -c 'make test-kevm-pyk -j`nproc`'
       - name: 'Test conformance'
-        run: docker exec -t -u user kevm-ci-concrete-${GITHUB_SHA} /bin/bash -c 'make test-conformance -j`nproc` TEST_CONCRETE_BACKEND=llvm'
+        run: docker exec -u user kevm-ci-concrete-${GITHUB_SHA} /bin/bash -c 'make test-conformance -j`nproc` TEST_CONCRETE_BACKEND=llvm'
       - name: 'Test llvm krun'
-        run: docker exec -t -u user kevm-ci-concrete-${GITHUB_SHA} /bin/bash -c 'make test-interactive-run TEST_CONCRETE_BACKEND=llvm'
+        run: docker exec -u user kevm-ci-concrete-${GITHUB_SHA} /bin/bash -c 'make test-interactive-run TEST_CONCRETE_BACKEND=llvm'
       - name: 'Test kast'
-        run: docker exec -t -u user kevm-ci-concrete-${GITHUB_SHA} /bin/bash -c 'make test-parse -j`nproc` TEST_CONCRETE_BACKEND=llvm'
+        run: docker exec -u user kevm-ci-concrete-${GITHUB_SHA} /bin/bash -c 'make test-parse -j`nproc` TEST_CONCRETE_BACKEND=llvm'
       - name: 'Test failure tests'
-        run: docker exec -t -u user kevm-ci-concrete-${GITHUB_SHA} /bin/bash -c 'make test-failure -j`nproc` TEST_CONCRETE_BACKEND=llvm'
+        run: docker exec -u user kevm-ci-concrete-${GITHUB_SHA} /bin/bash -c 'make test-failure -j`nproc` TEST_CONCRETE_BACKEND=llvm'
       - name: 'Test node'
-        run: docker exec -t -u user kevm-ci-concrete-${GITHUB_SHA} /bin/bash -c 'make test-node -j`nproc`'
+        run: docker exec -u user kevm-ci-concrete-${GITHUB_SHA} /bin/bash -c 'make test-node -j`nproc`'
       - name: 'Tear down Docker'
         if: always()
         run: |
@@ -57,11 +57,11 @@ jobs:
         with:
           tag: kevm-ci-java-${{ github.sha }}
       - name: 'Build kevm-pyk'
-        run: docker exec -t -u user kevm-ci-java-${GITHUB_SHA} /bin/bash -c 'make kevm-pyk venv -j2'
+        run: docker exec -u user kevm-ci-java-${GITHUB_SHA} /bin/bash -c 'make kevm-pyk venv -j2'
       - name: 'Build Definitions'
-        run: docker exec -t -u user kevm-ci-java-${GITHUB_SHA} /bin/bash -c 'make build-prove-java -j2'
+        run: docker exec -u user kevm-ci-java-${GITHUB_SHA} /bin/bash -c 'make build-prove-java -j2'
       - name: 'Prove Java'
-        run: docker exec -t -u user kevm-ci-java-${GITHUB_SHA} /bin/bash -c 'make test-prove -j3 TEST_SYMBOLIC_BACKEND=java'
+        run: docker exec -u user kevm-ci-java-${GITHUB_SHA} /bin/bash -c 'make test-prove -j3 TEST_SYMBOLIC_BACKEND=java'
       - name: 'Tear down Docker'
         if: always()
         run: |
@@ -81,11 +81,11 @@ jobs:
         with:
           tag: kevm-ci-haskell-${{ github.sha }}
       - name: 'Build kevm-pyk'
-        run: docker exec -t -u user kevm-ci-haskell-${GITHUB_SHA} /bin/bash -c 'make kevm-pyk venv -j2'
+        run: docker exec -u user kevm-ci-haskell-${GITHUB_SHA} /bin/bash -c 'make kevm-pyk venv -j2'
       - name: 'Build Definitions'
-        run: docker exec -t -u user kevm-ci-haskell-${GITHUB_SHA} /bin/bash -c 'make build-prove-haskell -j4'
+        run: docker exec -u user kevm-ci-haskell-${GITHUB_SHA} /bin/bash -c 'make build-prove-haskell -j4'
       - name: 'Prove Haskell'
-        run: docker exec -t -u user kevm-ci-haskell-${GITHUB_SHA} /bin/bash -c 'make test-prove -j6 TEST_SYMBOLIC_BACKEND=haskell'
+        run: docker exec -u user kevm-ci-haskell-${GITHUB_SHA} /bin/bash -c 'make test-prove -j6 TEST_SYMBOLIC_BACKEND=haskell'
       - name: 'Tear down Docker'
         if: always()
         run: |
@@ -105,11 +105,11 @@ jobs:
         with:
           tag: kevm-ci-foundry-${{ github.sha }}
       - name: 'Build kevm-pyk'
-        run: docker exec -t -u user kevm-ci-foundry-${GITHUB_SHA} /bin/bash -c 'make kevm-pyk venv -j2'
+        run: docker exec -u user kevm-ci-foundry-${GITHUB_SHA} /bin/bash -c 'make kevm-pyk venv -j2'
       - name: 'Build Foundry'
-        run: docker exec -t -u user kevm-ci-foundry-${GITHUB_SHA} /bin/bash -c 'make build-foundry -j2'
+        run: docker exec -u user kevm-ci-foundry-${GITHUB_SHA} /bin/bash -c 'make build-foundry -j2'
       - name: 'Test Foundry'
-        run: docker exec -t -u user kevm-ci-foundry-${GITHUB_SHA} /bin/bash -c 'make test-foundry -j2 FOUNDRY_PAR=4 KPROVE_OPTS=--retry'
+        run: docker exec -u user kevm-ci-foundry-${GITHUB_SHA} /bin/bash -c 'make test-foundry -j2 FOUNDRY_PAR=4 KPROVE_OPTS=--retry'
       - name: 'Tear down Docker'
         if: always()
         run: |

--- a/.github/workflows/test-pr.yml
+++ b/.github/workflows/test-pr.yml
@@ -23,21 +23,21 @@ jobs:
         with:
           tag: kevm-ci-concrete-${{ github.sha }}
       - name: 'Build kevm-pyk'
-        run: docker exec -t kevm-ci-concrete-${GITHUB_SHA} /bin/bash -c 'make kevm-pyk venv -j`nproc`'
+        run: docker exec -t -u user kevm-ci-concrete-${GITHUB_SHA} /bin/bash -c 'make kevm-pyk venv -j`nproc`'
       - name: 'Build kevm'
-        run: docker exec -t kevm-ci-concrete-${GITHUB_SHA} /bin/bash -c 'make build-llvm build-node -j`nproc` RELEASE=true'
+        run: docker exec -t -u user kevm-ci-concrete-${GITHUB_SHA} /bin/bash -c 'make build-llvm build-node -j`nproc` RELEASE=true'
       - name: 'Test kevm-pyk'
-        run: docker exec -t kevm-ci-concrete-${GITHUB_SHA} /bin/bash -c 'make test-kevm-pyk -j`nproc`'
+        run: docker exec -t -u user kevm-ci-concrete-${GITHUB_SHA} /bin/bash -c 'make test-kevm-pyk -j`nproc`'
       - name: 'Test conformance'
-        run: docker exec -t kevm-ci-concrete-${GITHUB_SHA} /bin/bash -c 'make test-conformance -j`nproc` TEST_CONCRETE_BACKEND=llvm'
+        run: docker exec -t -u user kevm-ci-concrete-${GITHUB_SHA} /bin/bash -c 'make test-conformance -j`nproc` TEST_CONCRETE_BACKEND=llvm'
       - name: 'Test llvm krun'
-        run: docker exec -t kevm-ci-concrete-${GITHUB_SHA} /bin/bash -c 'make test-interactive-run TEST_CONCRETE_BACKEND=llvm'
+        run: docker exec -t -u user kevm-ci-concrete-${GITHUB_SHA} /bin/bash -c 'make test-interactive-run TEST_CONCRETE_BACKEND=llvm'
       - name: 'Test kast'
-        run: docker exec -t kevm-ci-concrete-${GITHUB_SHA} /bin/bash -c 'make test-parse -j`nproc` TEST_CONCRETE_BACKEND=llvm'
+        run: docker exec -t -u user kevm-ci-concrete-${GITHUB_SHA} /bin/bash -c 'make test-parse -j`nproc` TEST_CONCRETE_BACKEND=llvm'
       - name: 'Test failure tests'
-        run: docker exec -t kevm-ci-concrete-${GITHUB_SHA} /bin/bash -c 'make test-failure -j`nproc` TEST_CONCRETE_BACKEND=llvm'
+        run: docker exec -t -u user kevm-ci-concrete-${GITHUB_SHA} /bin/bash -c 'make test-failure -j`nproc` TEST_CONCRETE_BACKEND=llvm'
       - name: 'Test node'
-        run: docker exec -t kevm-ci-concrete-${GITHUB_SHA} /bin/bash -c 'make test-node -j`nproc`'
+        run: docker exec -t -u user kevm-ci-concrete-${GITHUB_SHA} /bin/bash -c 'make test-node -j`nproc`'
       - name: 'Tear down Docker'
         if: always()
         run: |
@@ -57,11 +57,11 @@ jobs:
         with:
           tag: kevm-ci-java-${{ github.sha }}
       - name: 'Build kevm-pyk'
-        run: docker exec -t kevm-ci-java-${GITHUB_SHA} /bin/bash -c 'make kevm-pyk venv -j2'
+        run: docker exec -t -u user kevm-ci-java-${GITHUB_SHA} /bin/bash -c 'make kevm-pyk venv -j2'
       - name: 'Build Definitions'
-        run: docker exec -t kevm-ci-java-${GITHUB_SHA} /bin/bash -c 'make build-prove-java -j2'
+        run: docker exec -t -u user kevm-ci-java-${GITHUB_SHA} /bin/bash -c 'make build-prove-java -j2'
       - name: 'Prove Java'
-        run: docker exec -t kevm-ci-java-${GITHUB_SHA} /bin/bash -c 'make test-prove -j3 TEST_SYMBOLIC_BACKEND=java'
+        run: docker exec -t -u user kevm-ci-java-${GITHUB_SHA} /bin/bash -c 'make test-prove -j3 TEST_SYMBOLIC_BACKEND=java'
       - name: 'Tear down Docker'
         if: always()
         run: |
@@ -81,11 +81,11 @@ jobs:
         with:
           tag: kevm-ci-haskell-${{ github.sha }}
       - name: 'Build kevm-pyk'
-        run: docker exec -t kevm-ci-haskell-${GITHUB_SHA} /bin/bash -c 'make kevm-pyk venv -j2'
+        run: docker exec -t -u user kevm-ci-haskell-${GITHUB_SHA} /bin/bash -c 'make kevm-pyk venv -j2'
       - name: 'Build Definitions'
-        run: docker exec -t kevm-ci-haskell-${GITHUB_SHA} /bin/bash -c 'make build-prove-haskell -j4'
+        run: docker exec -t -u user kevm-ci-haskell-${GITHUB_SHA} /bin/bash -c 'make build-prove-haskell -j4'
       - name: 'Prove Haskell'
-        run: docker exec -t kevm-ci-haskell-${GITHUB_SHA} /bin/bash -c 'make test-prove -j6 TEST_SYMBOLIC_BACKEND=haskell'
+        run: docker exec -t -u user kevm-ci-haskell-${GITHUB_SHA} /bin/bash -c 'make test-prove -j6 TEST_SYMBOLIC_BACKEND=haskell'
       - name: 'Tear down Docker'
         if: always()
         run: |
@@ -105,11 +105,11 @@ jobs:
         with:
           tag: kevm-ci-foundry-${{ github.sha }}
       - name: 'Build kevm-pyk'
-        run: docker exec -t kevm-ci-foundry-${GITHUB_SHA} /bin/bash -c 'make kevm-pyk venv -j2'
+        run: docker exec -t -u user kevm-ci-foundry-${GITHUB_SHA} /bin/bash -c 'make kevm-pyk venv -j2'
       - name: 'Build Foundry'
-        run: docker exec -t kevm-ci-foundry-${GITHUB_SHA} /bin/bash -c 'make build-foundry -j2'
+        run: docker exec -t -u user kevm-ci-foundry-${GITHUB_SHA} /bin/bash -c 'make build-foundry -j2'
       - name: 'Test Foundry'
-        run: docker exec -t kevm-ci-foundry-${GITHUB_SHA} /bin/bash -c 'make test-foundry -j2 FOUNDRY_PAR=4 KPROVE_OPTS=--retry'
+        run: docker exec -t -u user kevm-ci-foundry-${GITHUB_SHA} /bin/bash -c 'make test-foundry -j2 FOUNDRY_PAR=4 KPROVE_OPTS=--retry'
       - name: 'Tear down Docker'
         if: always()
         run: |

--- a/.github/workflows/test-pr.yml
+++ b/.github/workflows/test-pr.yml
@@ -23,7 +23,7 @@ jobs:
         with:
           tag: kevm-ci-concrete-${{ github.sha }}
       - name: 'Build kevm-pyk'
-        run: docker exec -t -u user kevm-ci-concrete-${GITHUB_SHA} /bin/bash -c 'make kevm-pyk venv -j`nproc`'
+        run: docker exec -t -u user kevm-ci-concrete-${GITHUB_SHA} /bin/bash -c 'make kevm-pyk venv'
       - name: 'Build kevm'
         run: docker exec -t -u user kevm-ci-concrete-${GITHUB_SHA} /bin/bash -c 'make build-llvm build-node -j`nproc` RELEASE=true'
       - name: 'Test kevm-pyk'

